### PR TITLE
Add `return_samples` argument to `F.negative_sampling` and `L.NegativeSampling`

### PR DESCRIPTION
**Pull request by niboshi**
Refactored test of `L.NegativeSampling` as well.

### CODE REVIEW

File: chainer/functions/loss/negative_sampling.py [hunk 1]
```diff
@@ -314,8 +317,10 @@ def backward(self, indexes, grad_outputs):
         return ret
 
 
-def negative_sampling(x, t, W, sampler, sample_size, reduce='sum'):
-    """Negative sampling loss function.
+def negative_sampling(x, t, W, sampler, sample_size, reduce='sum', **kwargs):
```
Comment (beam2d) [inline chainer/functions/loss/negative_sampling.py]: Why is `return_samples` keyword only?

File: chainer/functions/loss/negative_sampling.py [hunk 2]
```diff
@@ -16,6 +17,7 @@ def _sigmoid_grad(x, y, gy):
 class NegativeSamplingFunction(function_node.FunctionNode):
 
     ignore_label = -1
+    _samples = None
```
Comment (beam2d) [inline chainer/functions/loss/negative_sampling.py]: Do we need to change the attribute name from `samples` to `_samples`?

File: chainer/functions/loss/negative_sampling.py [hunk 3]
```diff
@@ -373,8 +378,15 @@ def negative_sampling(x, t, W, sampler, sample_size, reduce='sum'):
 
     .. seealso:: :class:`~chainer.links.NegativeSampling`.
 
-    """
-    return (
-        NegativeSamplingFunction(sampler, sample_size, reduce)
-        .apply((x, t, W))
-    )[0]
+    """  # NOQA
```
Comment (beam2d) [inline chainer/functions/loss/negative_sampling.py]: Add a documentation of the `return_samples` argument.

File: tests/chainer_tests/links_tests/loss_tests/test_negative_sampling.py [hunk 4]
```diff
@@ -34,14 +30,49 @@ def setUp(self):
...
+
+        assert saved_samples[0] is not None
+        return out, saved_samples[0]
+
+    def check_forward(self, link, x_data, t_data):
         x = chainer.Variable(x_data)
         t = chainer.Variable(t_data)
-        y = self.link(x, t, reduce=self.reduce)
+        y, samples = self.call_link_and_return_samples(
+            lambda: link(x, t, reduce=self.reduce))
```
Comment (beam2d) [inline tests/chainer_tests/links_tests/loss_tests/test_negative_sampling.py]: How about adding `return_samples` argument to `L.NegativeSampling.forward`? I like such a consistency and we can avoid the patching trick in the tests.

File: chainer/functions/loss/negative_sampling.py (hunk 1 above)
Comment (niboshi) [reply chainer/functions/loss/negative_sampling.py]: To be conservative: We don't have to define fixed ordering. It's benefitial for both users and us.

* For users: Forces easier-to-understand keyword notation.
* For us: Easier to deprecate/add options in the future.

File: tests/chainer_tests/links_tests/loss_tests/test_negative_sampling.py (hunk 4 above)
Comment (niboshi) [reply tests/chainer_tests/links_tests/loss_tests/test_negative_sampling.py]: Added the option as suggested.

File: chainer/functions/loss/negative_sampling.py (hunk 3 above)
Comment (niboshi) [reply chainer/functions/loss/negative_sampling.py]: Added documentation.

File: chainer/functions/loss/negative_sampling.py (hunk 2 above)
Comment (niboshi) [reply chainer/functions/loss/negative_sampling.py]: Reverted to `samples`.

**Comment by niboshi** [comment]
PTAL

File: chainer/functions/loss/negative_sampling.py (hunk 1 above)
Comment (beam2d) [reply chainer/functions/loss/negative_sampling.py]: OK.

File: chainer/functions/loss/negative_sampling.py [hunk 5]
```diff
@@ -359,22 +364,37 @@ def negative_sampling(x, t, W, sampler, sample_size, reduce='sum'):
         sample_size (int): Number of samples.
         reduce (str): Reduction option. Its value must be either
             ``'sum'`` or ``'no'``. Otherwise, :class:`ValueError` is raised.
+        return_samples (bool):
+            If ``True``, the sample array taken by ``sampler`` is also
+            returned.
```
Comment (beam2d) [inline chainer/functions/loss/negative_sampling.py]: Could you add a description of "the sample array"? It is a (batch size, sample size+1)-array of integers whose first column is fixed to the ground truth labels and the other columns are drawn from the sampler.

File: chainer/functions/loss/negative_sampling.py [hunk 6]
```diff
@@ -359,22 +364,37 @@ def negative_sampling(x, t, W, sampler, sample_size, reduce='sum'):
...
-            above equation.
+        ~chainer.Variable or tuple:
+            If ``return_samples`` is ``False`` (default), the output
+            variable holding the loss value(s) calculated by the
+            above equation is returned.
             If ``reduce`` is ``'no'``, the output variable holds array
             whose shape is same as one of (hence both of) input variables.
             If it is ``'sum'``, the output variable holds a scalar value.
 
+            If ``return_samples`` is ``True``, a tuple of the output
```
Comment (beam2d) [inline chainer/functions/loss/negative_sampling.py]: How about putting this sentence right after the "False" case description, above the description of "reduce" option? I was first confused by the combinations of interleaved "if" sentences.

**Comment by beam2d** [comment]
Jenkins, test this please.

File: chainer/functions/loss/negative_sampling.py (hunk 5 above)
Comment (niboshi) [reply chainer/functions/loss/negative_sampling.py]: Added detailed description.

File: chainer/functions/loss/negative_sampling.py (hunk 6 above)
Comment (niboshi) [reply chainer/functions/loss/negative_sampling.py]: Fixed.

**Comment by niboshi** [comment]
PTAL

**Comment by beam2d** [comment]
LGTM

**Pull request merged**